pkg/privilege/privileges/cache.go

```diff
@@ -1686,6 +1686,148 @@ func (p *MySQLPrivilege) showGrants(ctx sessionctx.Context, user, host string, r
 	return gs
 }
 
+func (p *MySQLPrivilege) fetchColumnPrivileges(users, hosts []string) (rows [][]types.Datum) {
+	rows = make([][]types.Datum, 0, len(users))
+	// We use a 2-level map to record if a `username`@`hostname` is grantable on table/column level
+	// grantableUser records: username -> hostname -> grantable
+	grantableUser := make(map[string]map[string]struct{}, 4)
+	isGrantable := func(username, hostname string) string {
+		if hostNames, exists := grantableUser[username]; exists {
+			if _, exists = hostNames[hostname]; exists {
+				return "YES"
+			}
+		}
+		return "NO"
+	}
+
+	for i := range users {
+		userName, hostName := users[i], hosts[i]
+		columnsPrivRecords, exists := p.ColumnsPrivMap[userName]
+		if !exists {
+			continue
+		}
+
+		// need to get "grantable" information from mysql.tables_priv
+		if _, exists := grantableUser[userName]; !exists {
+			grantableUser[userName] = make(map[string]struct{}, 1)
+			if tablesPrivRecords, exists := p.TablesPrivMap[userName]; exists {
```

Comment:
Well, it's ok here in 8.5 that we still have full privilege data in-memory.
But in 9.0 later, access the whole data might cost a lot, considing if there is 1M users.
How to cherry-pick would be an issue later.